# Price data quality dashboard

## Imports and set up django environment

In [11]:
import os
import pandas as pd
import django
from django.db import connection
from IPython.core.display import display, HTML
os.chdir('..')

# Allows async calls to django ORM in Jupyter. Required.
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'rest.settings')
os.environ["DJANGO_ALLOW_ASYNC_UNSAFE"] = "true"
django.setup()

from pricedata import models





## Time ORM vs  Raw SQL
Retrieve data into dsataframe. We need:
* Symbol Name;
* Instrument Type Name;
* DataSource Name; and
* Time.

Where Retrieve Price Data is True and period is '1S'

## Django ORM QuertSet then convert to dataframe

In [32]:
%%timeit
price_data =  models.Candle.objects.filter(datasource_symbol__retrieve_price_data=True, period='1S')


df = pd.DataFrame(list(price_data.values('datasource_symbol__symbol__name',
                                                 'datasource_symbol__symbol__instrument_type',
                                                 'datasource_symbol__datasource__name', 'time')))

print(len(df.index))

2147012
2147012
2147012
2147012
2147012
2147012
2147012
2147012
25.4 s ± 1.45 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Raw sql straight into dataframe


In [61]:
%%timeit
sql = """SELECT	s.name AS symbol,
                s.instrument_type AS instrument_type,
                ds.name AS datasource,
                cdl.time AS time
        FROM public.pricedata_datasourcesymbol dss
            INNER JOIN pricedata_datasource ds ON dss.datasource_id = ds.id
            INNER JOIN pricedata_symbol s ON dss.symbol_id = s.id
            INNER JOIN pricedata_candle cdl ON cdl.datasource_symbol_id = dss.id  
        WHERE dss.retrieve_price_data = true
            AND cdl.period = %(period)s"""
    
df = pd.read_sql_query(sql=sql, con=connection, params={'period': '1S'})

print(len(df.index))

2240566


## % difference between results

In [55]:
100 - 22.2 / 25.4 * 100

12.5984251968504

## Results
### Django
Time: 25.4 s ± 1.45 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Rowcount: 2147012
    
### Raw
Time: 22.2 s ± 221 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Rowcount:214012

#### We have a winner. Raw SQL is 12.5% faster than Django
    


## Get the data using raw SQL, this time without %%timeit

In [2]:
sql = """SELECT	s.name AS symbol,
                s.instrument_type AS instrument_type,
                ds.name AS datasource,
                cdl.time AS time
        FROM public.pricedata_datasourcesymbol dss
            INNER JOIN pricedata_datasource ds ON dss.datasource_id = ds.id
            INNER JOIN pricedata_symbol s ON dss.symbol_id = s.id
            INNER JOIN pricedata_candle cdl ON cdl.datasource_symbol_id = dss.id  
        WHERE dss.retrieve_price_data = true
            AND cdl.period = %(period)s"""
    
df = pd.read_sql_query(sql=sql, con=connection, params={'period': '1S'})

df

symbol instrument_type datasource                      time
0        CADCHF           FOREX        MT5 2021-07-16 00:00:00+00:00
1        CADCHF           FOREX        MT5 2021-07-16 00:05:01+00:00
2        CADCHF           FOREX        MT5 2021-07-16 00:05:27+00:00
3        CADCHF           FOREX        MT5 2021-07-16 00:05:43+00:00
4        CADCHF           FOREX        MT5 2021-07-16 00:05:51+00:00
...         ...             ...        ...                       ...
4285989  USDCNH           FOREX        MT5 2021-07-20 10:25:13+00:00
4285990  USDCNH           FOREX        MT5 2021-07-20 10:25:14+00:00
4285991  USDCNH           FOREX        MT5 2021-07-20 10:25:15+00:00
4285992  USDCNH           FOREX        MT5 2021-07-20 10:25:16+00:00
4285993  USDCNH           FOREX        MT5 2021-07-20 10:25:17+00:00

[4285994 rows x 4 columns]

## Add another datasource for testing

In [136]:
nds = df.copy()
nds['datasource'] = 'NDS'
df = df.append(nds)

print(f"NDS has {len(df[df['datasource'] == 'NDS'].index)} rows. MT5 has {len(df[df['datasource'] == 'MT5'].index)} rows.")

df

NDS has 2473058 rows. MT5 has 2473058 rows.


symbol instrument_type datasource                      time
0        CADCHF           FOREX        MT5 2021-07-16 00:00:00+00:00
1        CADCHF           FOREX        MT5 2021-07-16 00:05:01+00:00
2        CADCHF           FOREX        MT5 2021-07-16 00:05:27+00:00
3        CADCHF           FOREX        MT5 2021-07-16 00:05:43+00:00
4        CADCHF           FOREX        MT5 2021-07-16 00:05:51+00:00
...         ...             ...        ...                       ...
2473053  USDCNH           FOREX        NDS 2021-07-19 12:25:07+00:00
2473054  USDCNH           FOREX        NDS 2021-07-19 12:25:08+00:00
2473055  USDCNH           FOREX        NDS 2021-07-19 12:25:10+00:00
2473056  USDCNH           FOREX        NDS 2021-07-19 12:25:11+00:00
2473057  USDCNH           FOREX        NDS 2021-07-19 12:25:12+00:00

[4946116 rows x 4 columns]

## Table showing num prices by symbol, by DS for each aggregarion period

In [82]:
# First group by to get min and max times
grouped = df.groupby(['symbol', 'instrument_type', 'datasource']).agg(first=('time', 'min'), last=('time', 'max'), count=('time', 'count'))

# Now get min, max and mean for each aggregation period for each datasource.
# 'minutes': 'T', 'hours': 'H', 'days': 'D', 'weeks': 'W', 'months': 'M'
aggs = {'minutes': 'T', 'hours': 'H', 'days': 'D', 'weeks': 'W', 'months': 'M'}

for key in aggs:
    # Get counts for aggregation period, then group by symbol, instrument type and datasource to get min, max and avg counts for aggregation period
    agg_period_ungrouped = df.groupby(['symbol', 'instrument_type', 'datasource', pd.Grouper(key='time', freq=aggs[key])]).agg(count=('time', 'count'))
    agg_period_grouped = agg_period_ungrouped.groupby(['symbol', 'instrument_type', 'datasource']).agg(min=('count', 'min'), max=('count', 'max'), avg=('count', 'median'))

    # Rename columns to include aggregation period key, then merge into original dataframe
    agg_period_grouped = agg_period_grouped.rename(columns={'min': f'{key}_min', 'max': f'{key}_max', 'avg': f'{key}_avg'})
    grouped = grouped.join(agg_period_grouped, on=['symbol', 'instrument_type', 'datasource'])
    
# Now unstack, so we show columns for each datasource
grouped = grouped.unstack()

# Create the same aggregations but across all datasources
agg_cols = set([x[0] for x in grouped.columns])
ds_cols = set([x[1] for x in grouped.columns])

for agg_col in agg_cols:
    if '_min' in agg_col or agg_col == 'first':
        grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].min(axis=1)
    elif '_max' in agg_col or agg_col == 'last':
        grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].max(axis=1)
    elif '_avg' in agg_col:
        grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].mean(axis=1)
        

# Flatten the columns and reset the index
grouped.columns = ['_'.join(col).strip() for col in grouped.columns]
grouped = grouped.reset_index()
                          
grouped


symbol instrument_type                 first_MT5  \
0   AUDCAD           FOREX 2021-07-16 00:00:00+00:00   
1   AUDCHF           FOREX 2021-07-16 00:00:00+00:00   
2   AUDJPY           FOREX 2021-07-16 00:00:00+00:00   
3   AUDNZD           FOREX 2021-07-16 00:00:00+00:00   
4   AUDUSD           FOREX 2021-07-16 00:00:00+00:00   
5   CADCHF           FOREX 2021-07-16 00:00:00+00:00   
6   CADJPY           FOREX 2021-07-16 00:00:00+00:00   
7   CHFJPY           FOREX 2021-07-16 00:00:00+00:00   
8   CHFSGD           FOREX 2021-07-16 00:00:00+00:00   
9   EURAUD           FOREX 2021-07-16 00:00:00+00:00   
10  EURCAD           FOREX 2021-07-16 00:00:00+00:00   
11  EURCHF           FOREX 2021-07-16 00:00:00+00:00   
12  EURDKK           FOREX 2021-07-16 00:00:00+00:00   
13  EURGBP           FOREX 2021-07-16 00:00:00+00:00   
14  EURHKD           FOREX 2021-07-16 00:00:00+00:00   
15  EURHUF           FOREX 2021-07-16 00:00:00+00:00   
16  EURJPY           FOREX 2021-07-16 00:00:00+00:00   
17  EURNOK           FOREX 2021-07-16 00:00:00+00:00   
18  EURNZD           FOREX 2021-07-16 00:00:00+00:00   
19  EURPLN           FOREX 2021-07-16 00:00:00+00:00   
20  EURRUB           FOREX 2021-07-16 10:00:00+00:00   
21  EURSEK           FOREX 2021-07-16 00:00:00+00:00   
22  EURSGD           FOREX 2021-07-16 00:00:00+00:00   
23  EURTRY           FOREX 2021-07-16 00:00:00+00:00   
24  EURUSD           FOREX 2021-07-16 00:00:00+00:00   
25  EURZAR           FOREX 2021-07-16 00:00:00+00:00   
26  GBPAUD           FOREX 2021-07-16 00:00:00+00:00   
27  GBPCAD           FOREX 2021-07-16 00:00:00+00:00   
28  GBPCHF           FOREX 2021-07-16 00:00:00+00:00   
29  GBPDKK           FOREX 2021-07-16 00:00:00+00:00   
30  GBPJPY           FOREX 2021-07-16 00:00:00+00:00   
31  GBPNOK           FOREX 2021-07-16 00:00:00+00:00   
32  GBPNZD           FOREX 2021-07-16 00:00:00+00:00   
33  GBPSEK           FOREX 2021-07-16 00:00:00+00:00   
34  GBPSGD           FOREX 2021-07-16 00:00:00+00:00   
35  GBPUSD           FOREX 2021-07-16 00:00:00+00:00   
36  NZDCAD           FOREX 2021-07-16 00:00:00+00:00   
37  NZDCHF           FOREX 2021-07-16 00:00:00+00:00   
38  NZDJPY           FOREX 2021-07-16 00:00:00+00:00   
39  NZDSGD           FOREX 2021-07-16 00:00:00+00:00   
40  NZDUSD           FOREX 2021-07-16 00:00:00+00:00   
41  SGDJPY           FOREX 2021-07-16 00:00:00+00:00   
42  USDCAD           FOREX 2021-07-16 00:00:00+00:00   
43  USDCHF           FOREX 2021-07-16 00:00:00+00:00   
44  USDCNH           FOREX 2021-07-16 00:00:00+00:00   
45  USDDKK           FOREX 2021-07-16 00:00:00+00:00   
46  USDHKD           FOREX 2021-07-16 00:00:00+00:00   
47  USDHUF           FOREX 2021-07-16 00:00:00+00:00   
48  USDJPY           FOREX 2021-07-16 00:00:00+00:00   
49  USDMXN           FOREX 2021-07-16 00:00:00+00:00   
50  USDNOK           FOREX 2021-07-16 00:00:00+00:00   
51  USDPLN           FOREX 2021-07-16 00:00:00+00:00   
52  USDRUB           FOREX 2021-07-16 10:00:00+00:00   
53  USDSEK           FOREX 2021-07-16 00:00:00+00:00   
54  USDSGD           FOREX 2021-07-16 00:00:00+00:00   
55  USDTRY           FOREX 2021-07-16 00:00:00+00:00   
56  USDZAR           FOREX 2021-07-16 00:00:00+00:00   

                    last_MT5  count_MT5  minutes_min_MT5  minutes_max_MT5  \
0  2021-07-19 21:24:47+00:00      75431                1               60   
1  2021-07-19 21:24:51+00:00      71641                1               60   
2  2021-07-19 21:24:52+00:00      76895                1               60   
3  2021-07-19 21:24:47+00:00      72478                1               60   
4  2021-07-19 21:24:47+00:00      70547                1               60   
5  2021-07-19 21:24:23+00:00      54384                1               59   
6  2021-07-19 21:24:26+00:00      64055                1               59   
7  2021-07-19 21:24:23+00:00      58093                1               59   
8  2021-07-19 21:24:58+00:00      41969                1      

## We will keep the full data table intact, and create a new one for display purposes. This will:
* Remove the DS agg columns; 
* Move the agg functions into the cells; 
* Rename the columns; and
* Reorder columns.

In [83]:
display = grouped

# Merge the agg functions for _all into cells for each aggregation period
for agg in ['minutes', 'hours', 'days', 'weeks', 'months']:
    # Get the min, max and avg columns and consolidate into a new column. Drop the existing ones.
    min_col = f'{agg}_min_all'
    max_col = f'{agg}_max_all'
    avg_col = f'{agg}_avg_all'
    display[agg] = 'min:' + display[min_col].astype(str) + ' max:' + display[max_col].astype(str) + ' avg:' + display[avg_col].astype(str)

# Rename the columns that we will be keeping
display = display.rename(
    columns={'symbol': 'Symbol', 'instrument_type': 'Instrument Type', 'first_all': 'First Price',
             'last_all': 'Last Price', 'minutes': 'Minutes', 'hours': 'Hours', 'days': 'Days', 'weeks': 'Weeks',
             'months': 'Months'})

# Reorder the columns, excluding any that we dont want to keep
display = display[
    ['Symbol', 'Instrument Type', 'First Price', 'Last Price', 'Minutes', 'Hours', 'Days', 'Weeks', 'Months']]

display

Symbol Instrument Type               First Price  \
0   AUDCAD           FOREX 2021-07-16 00:00:00+00:00   
1   AUDCHF           FOREX 2021-07-16 00:00:00+00:00   
2   AUDJPY           FOREX 2021-07-16 00:00:00+00:00   
3   AUDNZD           FOREX 2021-07-16 00:00:00+00:00   
4   AUDUSD           FOREX 2021-07-16 00:00:00+00:00   
5   CADCHF           FOREX 2021-07-16 00:00:00+00:00   
6   CADJPY           FOREX 2021-07-16 00:00:00+00:00   
7   CHFJPY           FOREX 2021-07-16 00:00:00+00:00   
8   CHFSGD           FOREX 2021-07-16 00:00:00+00:00   
9   EURAUD           FOREX 2021-07-16 00:00:00+00:00   
10  EURCAD           FOREX 2021-07-16 00:00:00+00:00   
11  EURCHF           FOREX 2021-07-16 00:00:00+00:00   
12  EURDKK           FOREX 2021-07-16 00:00:00+00:00   
13  EURGBP           FOREX 2021-07-16 00:00:00+00:00   
14  EURHKD           FOREX 2021-07-16 00:00:00+00:00   
15  EURHUF           FOREX 2021-07-16 00:00:00+00:00   
16  EURJPY           FOREX 2021-07-16 00:00:00+00:00   
17  EURNOK           FOREX 2021-07-16 00:00:00+00:00   
18  EURNZD           FOREX 2021-07-16 00:00:00+00:00   
19  EURPLN           FOREX 2021-07-16 00:00:00+00:00   
20  EURRUB           FOREX 2021-07-16 10:00:00+00:00   
21  EURSEK           FOREX 2021-07-16 00:00:00+00:00   
22  EURSGD           FOREX 2021-07-16 00:00:00+00:00   
23  EURTRY           FOREX 2021-07-16 00:00:00+00:00   
24  EURUSD           FOREX 2021-07-16 00:00:00+00:00   
25  EURZAR           FOREX 2021-07-16 00:00:00+00:00   
26  GBPAUD           FOREX 2021-07-16 00:00:00+00:00   
27  GBPCAD           FOREX 2021-07-16 00:00:00+00:00   
28  GBPCHF           FOREX 2021-07-16 00:00:00+00:00   
29  GBPDKK           FOREX 2021-07-16 00:00:00+00:00   
30  GBPJPY           FOREX 2021-07-16 00:00:00+00:00   
31  GBPNOK           FOREX 2021-07-16 00:00:00+00:00   
32  GBPNZD           FOREX 2021-07-16 00:00:00+00:00   
33  GBPSEK           FOREX 2021-07-16 00:00:00+00:00   
34  GBPSGD           FOREX 2021-07-16 00:00:00+00:00   
35  GBPUSD           FOREX 2021-07-16 00:00:00+00:00   
36  NZDCAD           FOREX 2021-07-16 00:00:00+00:00   
37  NZDCHF           FOREX 2021-07-16 00:00:00+00:00   
38  NZDJPY           FOREX 2021-07-16 00:00:00+00:00   
39  NZDSGD           FOREX 2021-07-16 00:00:00+00:00   
40  NZDUSD           FOREX 2021-07-16 00:00:00+00:00   
41  SGDJPY           FOREX 2021-07-16 00:00:00+00:00   
42  USDCAD           FOREX 2021-07-16 00:00:00+00:00   
43  USDCHF           FOREX 2021-07-16 00:00:00+00:00   
44  USDCNH           FOREX 2021-07-16 00:00:00+00:00   
45  USDDKK           FOREX 2021-07-16 00:00:00+00:00   
46  USDHKD           FOREX 2021-07-16 00:00:00+00:00   
47  USDHUF           FOREX 2021-07-16 00:00:00+00:00   
48  USDJPY           FOREX 2021-07-16 00:00:00+00:00   
49  USDMXN           FOREX 2021-07-16 00:00:00+00:00   
50  USDNOK           FOREX 2021-07-16 00:00:00+00:00   
51  USDPLN           FOREX 2021-07-16 00:00:00+00:00   
52  USDRUB           FOREX 2021-07-16 10:00:00+00:00   
53  USDSEK           FOREX 2021-07-16 00:00:00+00:00   
54  USDSGD           FOREX 2021-07-16 00:00:00+00:00   
55  USDTRY           FOREX 2021-07-16 00:00:00+00:00   
56  USDZAR           FOREX 2021-07-16 00:00:00+00:00   

                  Last Price                 Minutes  \
0  2021-07-19 21:24:47+00:00   min:1 max:60 avg:27.0   
1  2021-07-19 21:24:51+00:00   min:1 max:60 avg:26.0   
2  2021-07-19 21:24:52+00:00   min:1 max:60 avg:27.0   
3  2021-07-19 21:24:47+00:00   min:1 max:60 avg:25.0   
4  2021-07-19 21:24:47+00:00   min:1 max:60 avg:24.0   
5  2021-07-19 21:24:23+00:00   min:1 max:59 avg:18.0   
6  2021-07-19 21:24:26+00:00   min:1 max:59 avg:22.0   
7  2021-07-19 21:24:23+00:00   min:1 max:59 avg:20.0   
8  2021-07-19 21:24:58+00:00   min:1 max:56 avg:14.0   
9  2021-07-19 21:24:55+00:00   min:1 max:60 avg:36.0   
10 2021-07-19 21:24:26+00:00   min:1 max:60 avg:32.0   
11 2021-07-19 21:24:19+00:00   min:1 max:55 avg:16.0   
12 2021-07-19 21:25:15+00:0

### This is slow. Timeit shows: 21.9 s ± 60.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

## Tieing it all together

In [13]:
def retrieve_price_data(period: str) -> pd.DataFrame:
    """
    Retrieves the price data.
    
    :param period: The candle period that we are producing the dashboard for
    
    :return: DataFrame containing price data
    """
    sql = """SELECT	s.name AS symbol,
                s.instrument_type AS instrument_type,
                ds.name AS datasource,
                cdl.time AS time
        FROM public.pricedata_datasourcesymbol dss
            INNER JOIN pricedata_datasource ds ON dss.datasource_id = ds.id
            INNER JOIN pricedata_symbol s ON dss.symbol_id = s.id
            INNER JOIN pricedata_candle cdl ON cdl.datasource_symbol_id = dss.id  
        WHERE dss.retrieve_price_data = true
            AND cdl.period = %(period)s"""
    
    price_data = pd.read_sql_query(sql=sql, con=connection, params={'period': period})
    
    return price_data
    

def summarise_data(price_data: pd.DataFrame) -> pd.DataFrame:
    """
    Produces a data summary dashboard from provided data showing:
        * The datetime of the first and last candles for each datasource;
        * The number of candles for each datasource;
        * The minimum, maximum and average number of candles for each aggregation period for each datasource; and
        * The minimum, maximum and average number of candles for each aggregation period across all datasources.

    :param price_data: The candle period that we are producing the dashboard for
    """
    # First group by to get min and max times
    grouped = price_data.groupby(['symbol', 'instrument_type', 'datasource']).agg(first=('time', 'min'), last=('time', 'max'), count=('time', 'count'))

    # Now get min, max and mean for each aggregation period for each datasource.
    # 'minutes': 'T', 'hours': 'H', 'days': 'D', 'weeks': 'W', 'months': 'M'
    aggs = {'minutes': 'T', 'hours': 'H', 'days': 'D', 'weeks': 'W', 'months': 'M'}

    for key in aggs:
        # Get counts for aggregation period, then group by symbol, instrument type and datasource to get min, max and avg counts for aggregation period
        agg_period_ungrouped = price_data.groupby(['symbol', 'instrument_type', 'datasource', pd.Grouper(key='time', freq=aggs[key])]).agg(count=('time', 'count'))
        agg_period_grouped = agg_period_ungrouped.groupby(['symbol', 'instrument_type', 'datasource']).agg(min=('count', 'min'), max=('count', 'max'), avg=('count', 'median'))

        # Rename columns to include aggregation period key, then merge into original dataframe
        agg_period_grouped = agg_period_grouped.rename(columns={'min': f'{key}_min', 'max': f'{key}_max', 'avg': f'{key}_avg'})
        grouped = grouped.join(agg_period_grouped, on=['symbol', 'instrument_type', 'datasource'])

    # Now unstack, so we show columns for each datasource
    grouped = grouped.unstack()

    # Create the same aggregations but across all datasources
    agg_cols = set([x[0] for x in grouped.columns])
    ds_cols = set([x[1] for x in grouped.columns])

    for agg_col in agg_cols:
        if '_min' in agg_col or agg_col == 'first':
            grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].min(axis=1)
        elif '_max' in agg_col or agg_col == 'last':
            grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].max(axis=1)
        elif '_avg' in agg_col:
            grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].mean(axis=1)


    # Flatten the columns and reset the index
    grouped.columns = ['_'.join(col).strip() for col in grouped.columns]
    grouped = grouped.reset_index()
    
    return grouped

def format_data(summarised_data: pd.DataFrame) -> str:
    """
    Returns the HTML for the provided summary data. Formats as required for display.
    
    :param summary_data: A dataframe containing the summary data
    
    :return: html as string
    """
    
    display = summarised_data

    # Merge the agg functions for _all into cells for each aggregation period
    for agg in ['minutes', 'hours', 'days', 'weeks', 'months']:
        # Get the min, max and avg columns and consolidate into a new column. Drop the existing ones.
        min_col = f'{agg}_min_all'
        max_col = f'{agg}_max_all'
        avg_col = f'{agg}_avg_all'
        display[agg] = 'min:' + display[min_col].astype(str) + ' max:' + display[max_col].astype(str) + ' avg:' + display[avg_col].astype(str)

    # Rename the columns that we will be keeping
    display = display.rename(
        columns={'symbol': 'Symbol', 'instrument_type': 'Instrument Type', 'first_all': 'First Price',
                 'last_all': 'Last Price', 'minutes': 'Minutes', 'hours': 'Hours', 'days': 'Days', 'weeks': 'Weeks',
                 'months': 'Months'})

    # Reorder the columns, excluding any that we dont want to keep
    display = display[
        ['Symbol', 'Instrument Type', 'First Price', 'Last Price', 'Minutes', 'Hours', 'Days', 'Weeks', 'Months']]
    
    # Convert to HTML
    table_styles = 'table table-striped thead-dark'
    html = display.to_html(classes=table_styles)

    return html


## Run it and display output

In [14]:
data = retrieve_price_data('1S')
summarised_data = summarise_data(data)
formatted_data = format_data(summarised_data)

display(HTML(formatted_data))

In [28]:
import functools


def decorator_factory(request):
    def mydecorator(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            # Prints the caller signiture
            caller = f"{func.__name__}_{','.join([f'{x}' for x in args])}_" \
                          f"{','.join([f'{x}={kwargs[x]}' for x in kwargs.keys()])}"

            print(f'caller: {caller}')
        return wrapper
    return mydecorator


@decorator_factory('req')
def myfunc(myfuncparam):
    print('myfunc')


myfunc('myfuncparam')

caller: myfunc_myfuncparam_
